### ISSUE TITLE
coefficients in Bernoulli DE

### ISSUE THREAD

**Issue by poluyan** [sympy/sympy#22462]
Hello, everyone! 

I've ran into problems using Bernoulli equation with certain coefficients. I wanted to solve this equation with `dsolve(de, f(x))`
```
de = Eq(f(x).diff(x), 20*f(x)**2 - 500*f(x)/7200)
```
but got stuck on a long executing time. I checked the reduced coefficients
```
de = Eq(f(x).diff(x), 2*f(x)**2 - 5*f(x)/7)
```
and got this answer:
```
Eq(x - 7*log(f(x) - 5/14)/5 + 7*log(f(x))/5, C1)
```
By using `classify_ode` I found out that equations aren't completely classified. They aren't classified as 'Bernoulli' so I can't use a hint to point the type manually. I checked all available coefficients from -2 to 2 and it turns out that equations based on some of them aren't classified as Bernoulli equations. 
```
from sympy import *
f = Function("f")
x = symbols("x")
for i in range(-2,3,1):
    for j in range(-2,3,1):
        if i != 0 and j != 0:
            de = Eq(f(x).diff(x), i*f(x) + j*f(x)**2)
            if classify_ode(de).count('Bernoulli') == 0:
                print(de) 
# gives
#Eq(Derivative(f(x), x), -2*f(x)**2 - 2*f(x))
#Eq(Derivative(f(x), x), -2*f(x)**2 - f(x))
#Eq(Derivative(f(x), x), -2*f(x)**2 + f(x))
#Eq(Derivative(f(x), x), -2*f(x)**2 + 2*f(x))
```
Using `dsolve` leads to the correct answers for this four equations. However changing `-2*f(x)**2` to `-0.5*f(x)**2` leads to incomplete answers.

I am running SymPy 1.9. I checked live version with SymPy 1.5.1 and have similar results, except that the type is classified as Bernoulli and I can manually specify it with `hint='Bernoulli'` and get the right answers.

Any suggestions to solve the initial equation with SymPy 1.9?

 Thanks in advance

**Comment by oscarbenjamin** [comment]
I think there's a problem with the factorable solver.

I just tried with sympy 1.9 and I can specify the Bernoulli hint to get a solution:
```python
In [1]: de = Eq(f(x).diff(x), 20*f(x)**2 - 500*f(x)/7200)

In [2]: dsolve(de, hint='Bernoulli')
Out[2]: 
             1      
f(x) = ─────────────
           5⋅x      
           ───      
            72      
       C₁⋅ℯ    + 288

In [3]: sol = dsolve(de, hint='Bernoulli')

In [4]: checkodesol(de, sol)
Out[4]: (True, 0)
```
Without passing that hint the slowness comes down to this being slow:
```python
In [11]: eq = Eq(x - 72*log(z - 1/S(288)*1)/5 + 72*log(z)/5, y)

In [12]: eq
Out[12]: 
    72⋅log(z)   72⋅log(z - 1/288)    
x + ───────── - ───────────────── = y
        5               5            

In [13]: solve(eq, z)
^C---------------------------------------------------------------------------
KeyboardInterrupt
```
I'm not sure why that is slow but it ends up trying to compute the roots of a huge polynomial:
```python
Poly((1191952966037002531340318188592343200639306252477548716819283242528558379842401773317091519794202199699847970068712545348731017608260563475529278558677252206818987406223113977856*exp(5*x) - 1191952966037002531340318188592343200639306252477548716819283242528558379842401773317091519794202199699847970068712545348731017608260563475529278558677252206818987406223113977856*exp(5*y))*exp(-5*x)*z**72 + 297988241509250632835079547148085800159826563119387179204820810632139594960600443329272879948550549924961992517178136337182754402065140868882319639669313051704746851555778494464*exp(-5*x)*exp(5*y)*z**71 - 36731189491591657866824041401934187172478621495618905770038676310558873684379568535379122354769251813667190049860499444340235351643446183490702594473127129637217059827187974144*exp(-5*x)*exp(5*y)*z**70 + 2975906556031731540136207058027075349622110537839494680442948312198056895725196524857104820409545864533221416076660834610898697471112538014293034274443170225237493273036062720*exp(-5*x)*exp(5*y)*z**69 - 178244403095650587039408235246413367295074329089344733464030758282696116150207083520087007472446757511104407733758331239715286567280178058147759865396335716615787357499555840*exp(-5*x)*exp(5*y)*z**68 + 8417096812850166610194277775525075677822954429219056858023674696682872151537556721781886463976652438024374809649698975208777421232675074968088660310382519951301069659701248*exp(-5*x)*exp(5*y)*z**67 - 326357341701945117409153131342696800008181682151433338823834609188514140134847396041311570073168815594695088105630689432284772698257656263230289491201174095334011381481472*exp(-5*x)*exp(5*y)*z**66 + 10684317734289869915180608466576383333601186022814781925780299705576355778224170703733414496443026701016803479648623761175989582383435175284324953580990818597244420227072*exp(-5*x)*exp(5*y)*z**65 - 301423894413559698127925152051851092310797348733923969260294913568777398257192315860534697165276360922783084278281486317899011655782676386059514749463716670495176785920*exp(-5*x)*exp(5*y)*z**64 + 7442565294161967855010497581527187464464132067504295537291232433796972796473884342235424621364848417846495908105715711553062016192164849038506537023795473345559920640*exp(-5*x)*exp(5*y)*z**63 - 162806115809793046828354634595907225785152888976656464878245709489308779922866219986399913592356059140392097989812531190223231604203606072717330497395525979434123264*exp(-5*x)*exp(5*y)*z**62 + 3186230801833070992221586914440103534936704266588605057591929920560967283843972739632826591769594591762724139952139183647045568011560472382725533724281127122763776*exp(-5*x)*exp(5*y)*z**61 - 56238448759206403508540741256031920032158263964671559176246448250642073007662713286343293431118423060626786035034864063214635315018862504440468043165841653497856*exp(-5*x)*exp(5*y)*z**60 + 901257191653948774175332391923588462053818332767172422696257183503879375122799892409347651139718318279275417228122821525875565945815104237828013512273103421440*exp(-5*x)*exp(5*y)*z**59 - 13188039262793397241156897600072351007235932944757731383700191921311726967322716679601069299911552772439794051701202001494706942163465066972185713597250273280*exp(-5*x)*exp(5*y)*z**58 + 177061638250466907404421310371341749634186136758321393577456280425018556505721659124273615600664365926275012731173545390438195056824299510274715599222341632*exp(-5*x)*exp(5*y)*z**57 - 2190215577316973463986982354853836746777042056255277654929472218799057665109838231354947067976968068098453933523631095324430798228946413213033591396630528*exp(-5*x)*exp(5*y)*z**56 + 25051485361468650731877249156824930110194925480044025464879584201950005973478541861902989339605843916158786821348721678547411090853962242632737156497408*exp(-5*x)*exp(5*y)*z**55 - 265785434969285453366753222149955855721589680054479436838035712019145510906890393982381252638565087845048857093784662870391128471637330892129734492160*exp(-5*x)*exp(5*y)*z**54 + 2622882581933738026645591008058774891989371842642889179322720842294199120791681519562972887880576524786666352899190752010438767812210502224964485120*exp(-5*x)*exp(5*y)*z**53 - 24134162646265297814620889483874143971429983968762695573629202194720929410062347315423188030845582606543978594384914905651606717716520246167207936*exp(-5*x)*exp(5*y)*z**52 + 207502721164979412427296007467171872770231343646769207974325151145087356038895843320437463228169691722931032888230088474517782625869552381067264*exp(-5*x)*exp(5*y)*z**51 - 1670239706346898679575772787377803900139172747156759723278185402209509968116112375211854580908562859512228957907155068213448060908987874279424*exp(-5*x)*exp(5*y)*z**50 + 12607485706120913946073164155931490792113320857161531727643307685760189976721862735596728418693862164192549501110771952094263744784026828800*exp(-5*x)*exp(5*y)*z**49 - 89375983738415043888539502841528218867701493345039793786823217101019865286367371823530048107060076106110376960999396072427506292595097600*exp(-5*x)*exp(5*y)*z**48 + 595839891589433625923596685610188125784676622300265291912154780673465768575782478823533654047067174040735846406662640482850041950633984*exp(-5*x)*exp(5*y)*z**47 - 3739913849452908709723430051239161580112152944459464305538364675701507895708036392188312198212093640480046044486263902603358970576896*exp(-5*x)*exp(5*y)*z**46 + 22123975961269778889824817689943599882350699002718024441199173750291841975639104171895879773374010733292453452465038518486948642816*exp(-5*x)*exp(5*y)*z**45 - 123459687283871533983397420144774552914903454256238975676334674945825011024771786673526114806774613467033780426702223875485204480*exp(-5*x)*exp(5*y)*z**44 + 650410229943767659874220125283773985662805554031910312471111793296970843521307305272407692947567408111768000332243516585410560*exp(-5*x)*exp(5*y)*z**43 - 3236995357359028862799938123518782567534796160112516601418727674973350262895395153554806805178865572778475001653526760783872*exp(-5*x)*exp(5*y)*z**42 + 15227800740264248682526590634832982508564229247841139926028960836568180000179950319142236314685523527855729174445354385408*exp(-5*x)*exp(5*y)*z**41 - 67745207286331835501691646704443607080201106679848821285502104416156182726495004675003438465940371597448447933187883008*exp(-5*x)*exp(5*y)*z**40 + 285122926289275401943146661213988245287041694780508507093864075825573159623295474221394943038469577430338585577390080*exp(-5*x)*exp(5*y)*z**39 - 1135599890245275804307875795276301222037849887299819421636100792197442118597684180416095054993904566971324023439360*exp(-5*x)*exp(5*y)*z**38 + 4281031332273857198779690498065421273555386479900112899024983145188769891538888775378136120016703724693483421696*exp(-5*x)*exp(5*y)*z**37 - 15277600240560640080521657834531306628235850670939831912029743091433688849048889341144968792497881733570494464*exp(-5*x)*exp(5*y)*z**36 + 51613514326218378650411006197740900771067063077499432135235618552140840706246247774138408082763113964765184*exp(-5*x)*exp(5*y)*z**35 - 165065149983337285523061514704032486018580702459108198531912157284807147726481969306911941054158350581760*exp(-5*x)*exp(5*y)*z**34 + 499663025234461156319808716162491499700119647757271968490474835085776622391416217631321758889012101120*exp(-5*x)*exp(5*y)*z**33 - 1431326374369550187374452051507137108515967740971351993071672704672797616225411040089723788484149248*exp(-5*x)*exp(5*y)*z**32 + 3878933263874119748982254882133162895707229650328867189896131990983191371884582764470796174753792*exp(-5*x)*exp(5*y)*z**31 - 9941049204703845256154919092768522632847562761259497593153116048320017570140713103389110566912*exp(-5*x)*exp(5*y)*z**30 + 24081999042402725911228001678218320331510568704601496107444564070542678222240099572163543040*exp(-5*x)*exp(5*y)*z**29 - 55111898061054218073359536668902406061695588102386630927706151992245712471982551104225280*exp(-5*x)*exp(5*y)*z**28 + 119068915564006026701702702679727420503663307628613091510476254304234563982678351151104*exp(-5*x)*exp(5*y)*z**27 - 242667626828816630506187573396183601569965980221358202806677148717869355942958596096*exp(-5*x)*exp(5*y)*z**26 + 466116895504523669707511591925293560935218342623767233523463790385978372821876736*exp(-5*x)*exp(5*y)*z**25 - 842948668085437770738410720351008320557035486515782757384736310738531490201600*exp(-5*x)*exp(5*y)*z**24 + 1433586170213329542072127075426884898906522936251331220042068555677774643200*exp(-5*x)*exp(5*y)*z**23 - 2289755688535179129698536301029052269086807467623654032011637276429778944*exp(-5*x)*exp(5*y)*z**22 + 3429644958317942596226021148055497679732418592573555875834424025153536*exp(-5*x)*exp(5*y)*z**21 - 4809197657897756044387449526520128957958119020035034281017822150656*exp(-5*x)*exp(5*y)*z**20 + 6301359614645906766755043928878575678666298506335212632360878080*exp(-5*x)*exp(5*y)*z**19 - 7698420311102895355474912207349082940757437732791219136757760*exp(-5*x)*exp(5*y)*z**18 + 8748204898980562903948763871987594250860724696353658109952*exp(-5*x)*exp(5*y)*z**17 - 9221198120205206433973771442447240963828888878845001728*exp(-5*x)*exp(5*y)*z**16 + 8987522534313066699779504329870605227903400466710528*exp(-5*x)*exp(5*y)*z**15 - 8070691930956417654256020411162540614137392660480*exp(-5*x)*exp(5*y)*z**14 + 6649581393207971231143142994131095138766684160*exp(-5*x)*exp(5*y)*z**13 - 5002578594427293171577596002529180370599936*exp(-5*x)*exp(5*y)*z**12 + 3417061881439407904083057378776762548224*exp(-5*x)*exp(5*y)*z**11 - 2105044841836552808294894218556473344*exp(-5*x)*exp(5*y)*z**10 + 1160187853745895507217203603701760*exp(-5*x)*exp(5*y)*z**9 - 566497975461863040633400197120*exp(-5*x)*exp(5*y)*z**8 + 242093151906779077193760768*exp(-5*x)*exp(5*y)*z**7 - 89154675049844988444672*exp(-5*x)*exp(5*y)*z**6 + 27722224828931899392*exp(-5*x)*exp(5*y)*z**5 - 7077773904445440*exp(-5*x)*exp(5*y)*z**4 + 1424672686080*exp(-5*x)*exp(5*y)*z**3 - 212004864*exp(-5*x)*exp(5*y)*z**2 + 20736*exp(-5*x)*exp(5*y)*z - exp(-5*x)*exp(5*y), z, domain='ZZ(exp(x),exp(y))')
```

**Comment by poluyan** [comment]
Ouch, sorry. I've lost one minus during lots of experiments. The initial DE should be like this:
```
from sympy import *
f = Function("f")
x = symbols("x")
de = Eq(f(x).diff(x), -20*f(x)**2 - 500*f(x)/7200)
classify_ode(de)
('separable',
 '1st_exact',
 '1st_rational_riccati',
 '1st_power_series',
 'lie_group',
 'separable_Integral',
 '1st_exact_Integral')
```


There is no option to use the hint and solve it as Bernoulli. Similar is this one
```
de = Eq(f(x).diff(x), -2*f(x)**2 - 5*f(x)/7)
classify_ode(de)
('separable',
 '1st_exact',
 '1st_rational_riccati',
 '1st_power_series',
 'lie_group',
 'separable_Integral',
 '1st_exact_Integral')
```
the result of `dsolve(de, f(x))` is
```
s = dsolve(de, f(x))
print(s)
checkodesol(de, s)
#Eq(x - 7*log(f(x) + 5/14)/5 + 7*log(f(x))/5, C1)
#(True, 0)
```

**Comment by oscarbenjamin** [comment]
Okay I guess that there's a bug in the Bernoulli matching code as well as the problem with solve being slow.

**Comment by mostlyaman** [comment]
>I checked all available coefficients from -2 to 2 and it turns out that equations based on some of them aren't classified as Bernoulli equations.

I have testing the classify_ode function and I have found out that for a differential equation `de = Eq(f(x).diff(x), i*f(x)+j*f(x)**2)`, for any value of j <= 2 causes the failure of classification as a Bernoulli equation.

```python
>>> from sympy import *
>>> init_printing()
>>> x,f=symbols('x'),Function('f')
>>> for i in range(-10,10,1):
...     for j in range(-10,10,1):
...             de = Eq(f(x).diff(x), i*f(x)+j*f(x)**2)
...             if classify_ode(de).count('Bernoulli') == 0:
...                     print(i,j)
... 
-10 -10
-10 -9
-10 -8
-10 -7
-10 -6
-10 -5
-10 -4
-10 -3
-10 -2
-9 -10
-9 -9
-9 -8
-9 -7
-9 -6
-9 -5
-9 -4
..Skipped a lot a lines..
9 -4
9 -3
9 -2
```

I would be working on this issue. I have managed to find the  source of the problem. `_matches_commutative` method of class AssocOp fails to match the general bernoulli equation with these cases.

**Comment by mostlyaman** [comment]
I have found the exact cause of the bug. I will try to explain it here.
```python
>>> x,f = symbols('x'),Function('f')
>>> pattern = -Wild('Q')*f(x)**Wild('n')
>>> pattern
    n   
-Q⋅f (x)
>>> eq = 5*f(x)**2
>>> eq
   2   
5⋅f (x)
>>> print(pattern.matches(eq))
None
```

So, _matches_commutative method, which is where the actual matching of expression happens fails to return any output here. 
and the problem happens in the following code. Here `self` and `expr` are equivalent to pattern and eq from previous code example.
https://github.com/sympy/sympy/blob/7a0fbd8168fd091abf14b65badf1c7c7a7a810f9/sympy/core/operations.py#L255-L258

To remove the negation from the inbuilt bernoulli pattern, it adds a negation to eq and stores it in newexpr variable. But now newexpr has more operations than expr and thus it returns none. 

```python
>>> newexpr
    2   
-5⋅f (x)
>>> expr
   2   
5⋅f (x)
>>> newexpr.count_ops() > expr.count_ops()   
True                                                                    #and thus code return None 
>>> print(newexpr.count_ops(visual=True), '\n', expr.count_ops(visual=True))
F + MUL + NEG + POW 
 F + MUL + POW
```
This makes sense now why only bernoulli differential equations in the form of `f(x).diff(x) = i*f(x)+j*(f(x)**2)` fails to match the general equation only if j <= -2 as this adds the mentioned negation as well as keep the MUL operation. In fact, it does not need f(x)**2 it can be any power of f(x).


A very simple solution could be that we can change the default General Bernoulli Equation in sympy from `f(x).diff(x) + P_*f(x) - Q_*f(x)**n` to `f(x).diff(x) + P_*f(x) + Q_*f(x)**n` since the bug arises when the code tries to compensate for the negation in `Q_*f(x)**n`. Since the function used to retrieve the general equation( _equation method of Bernoulli class in single.py) is for internal use, I doubt it will break compatibility, although we may need to track down all the places where this method is used apart from this and change the sign suitably.

Apart from the mentioned concern, this solution works correctly and now classify_ode can detect these equations as Bernoulli.
```python
>>> for i in range(-10,11):
...     for j in range(-10,11):
...             de = Eq(f(x).diff(x),i*f(x)+j*f(x)**2)
...             if classify_ode(de).count('Bernoulli') == 1:
...                     print(i,j)
... 

-10 -10
-10 -9
-10 -8
-10 -7
-10 -6
-10 -5
-10 -4
-10 -3
-10 -2
-10 -1
-10 0
-10 1
-10 2
-10 3
-10 4  ..all the possible cases have been detected as Bernoulli
```
```python
>>> de
d                2      5⋅f(x)
──(f(x)) = - 20⋅f (x) - ──────
dx                        72  
>>> classify_ode(de)
(factorable, separable, 1st_exact, Bernoulli, 1st_rational_riccati, 1st_power_
series, lie_group, separable_Integral, 1st_exact_Integral, Bernoulli_Integral)
>>> dsolve(de, hint = 'Bernoulli')
             1      
f(x) = ─────────────
           5⋅x      
           ───      
            72      
       C₁⋅ℯ    - 288
```

However, it still gets stuck on this equation as it tries to compute the roots as described by oscarbenjamin.
Is this solution acceptable? To be more clear, Im attaching the git diff for the exact changes.
```
diff --git a/sympy/solvers/ode/single.py b/sympy/solvers/ode/single.py
index 46c3146af9..4251a286a2 100644
--- a/sympy/solvers/ode/single.py
+++ b/sympy/solvers/ode/single.py
@@ -830,7 +830,7 @@ def _wilds(self, f, x, order):
 
     def _equation(self, fx, x, order):
         P, Q, n = self.wilds()
-        return fx.diff(x) + P*fx - Q*fx**n
+        return fx.diff(x) + P*fx + Q*fx**n
 
     def _get_general_solution(self, *, simplify_flag: bool = True):
         P, Q, n = self.wilds_match()
@@ -839,11 +839,11 @@ def _get_general_solution(self, *, simplify_flag: bool = True):
         (C1,) = self.ode_problem.get_numbered_constants(num=1)
         if n==1:
             gensol = Eq(log(fx), (
-            C1 + Integral((-P + Q), x)
+            C1 + Integral((-P - Q), x)
         ))
         else:
             gensol = Eq(fx**(1-n), (
-                (C1 - (n - 1) * Integral(Q*exp(-n*Integral(P, x))
+                (C1 - (n - 1) * Integral(-Q*exp(-n*Integral(P, x))
                             * exp(Integral(P, x)), x)
                 ) * exp(-(1 - n)*Integral(P, x)))
             )
```

**Comment by oscarbenjamin** [comment]
I don't see why `count_ops` is being checked in the matching code. Is it possible to just remove that part?

**Comment by mostlyaman** [comment]
I dont know the exact reason but it is meant to ensure that the newexpr created after taking into account the constant in wild part does not have more operations than the previous expr. If so, then return None. I may try running tests after removing it and see if something fails.

**Comment by mostlyaman** [comment]
Some tests fail in if we remove count_ops. Particularly, 
https://github.com/sympy/sympy/blob/26110e2f95483393fd928072320734ef9cfa45c0/sympy/solvers/ode/tests/test_ode.py#L744-L751

After removing count_ops, it gives
```python
>>> classify_ode(eq)
(factorable, 1st_exact, 1st_linear, Bernoulli, almost_linear, 1st_power_series
, lie_group, nth_linear_constant_coeff_undetermined_coefficients, nth_linear_c
onstant_coeff_variation_of_parameters, 1st_exact_Integral, 1st_linear_Integral
, Bernoulli_Integral, almost_linear_Integral, nth_linear_constant_coeff_variat
ion_of_parameters_Integral)
```
I think the equation being tested is solvable by a Bernoulli equation, so I dont know why it was not included in the test. Should the test be modified? The answer given by Bernoulli contains the correct answer but also contains a wrong answer for A = -1.
```python>>> dsolve(eq,hint='Bernoulli')
       ⎛     ⎛⎧                    x            ⎞⎞    
       ⎜     ⎜⎪(-A⋅x + A - x - 1)⋅ℯ             ⎟⎟    
       ⎜     ⎜⎪─────────────────────  for A ≠ -1⎟⎟    
       ⎜     ⎜⎪        A + 1                    ⎟⎟  -x
f(x) = ⎜C₁ + ⎜⎨                                 ⎟⎟⋅ℯ  
       ⎜     ⎜⎪       2                         ⎟⎟    
       ⎜     ⎜⎪      x     2⋅x                  ⎟⎟    
       ⎜     ⎜⎪    - ── - ─────       otherwise ⎟⎟    
       ⎝     ⎝⎩      2    A + 1                 ⎠⎠  
```

**Comment by oscarbenjamin** [comment]
There are two separate issues:

1. Should the call to match that happens to be used by the Bernoulli solver in dsolve succeed according to the expected behaviour of match?
2. Should the Bernoulli solver be improved to handle this case correctly or to not match it in some way?

We need to investigate what the call to match used by Bernoulli is first without considering the validity of what dsolve returns which can be considered after. The questions are:

1. What is the expression being matched?
2. What is the wild card pattern it is being matched against?
3. Should the call to match succeed?
4. Does the change in the matching code make the behaviour of match correct in each case that we consider?

We should write a small self-contained code demonstration of this call to match so that it can be studied independently of `dsolve`.

**Comment by mostlyaman** [comment]
> Should the call to match that happens to be used by the Bernoulli solver in dsolve succeed according to the expected behaviour of match? 

Yes, for this particular case the Bernoulli solver does return the correct integral. It is the integration code that is adding a wrong Piecewise. Here `expr` is the expression returned by the Bernoulli solver, which seems to be the correct unevaluated integral. This get integrated in the `odesimp` function which creates the wrong result.
```python
(Pdb) pprint(expr)
       ⎛     ⌠             ⎞         
       ⎜     ⎮    ⌠        ⎟         
       ⎜     ⎮    ⎮ 1 dx   ⎟   ⌠     
       ⎜     ⎮    ⌡        ⎟  -⎮ 1 dx
       ⎜     ⎮ 2⋅ℯ         ⎟   ⌡     
f(x) = ⎜C₁ + ⎮ ───────── dx⎟⋅ℯ       
       ⎜     ⎮   A - 1     ⎟         
       ⎝     ⌡             ⎠             
(Pdb) pprint(expr.doit())
       ⎛     ⎛⎧    x           ⎞⎞    
       ⎜     ⎜⎪ 2⋅ℯ            ⎟⎟    
       ⎜     ⎜⎪─────  for A ≠ 1⎟⎟    
       ⎜     ⎜⎪A - 1           ⎟⎟  -x
f(x) = ⎜C₁ + ⎜⎨                ⎟⎟⋅ℯ  
       ⎜     ⎜⎪ 2⋅x            ⎟⎟    
       ⎜     ⎜⎪─────  otherwise⎟⎟    
       ⎜     ⎜⎪A - 1           ⎟⎟    
       ⎝     ⎝⎩                ⎠⎠ 
```
>Should the Bernoulli solver be improved to handle this case correctly or to not match it in some way?

No. as explained above, the matching of the expression is correct. Thus, instead of the Bernoulli solver, the integrate functions must be improved to handle cases like this.

**Comment by oscarbenjamin** [comment]
Okay well let's get match fixed then and open a separate issue for integrals.

**Comment by mostlyaman** [comment]
>Should the call to match succeed?

I think that this case should be classified as Bernoulli because it is solvable by that method and matches the general Bernoulli equation, being a linear DE. The test for this should be updated to include Bernoulli and the integrate functions being used need to be modified to handle this case. 
>Does the change in the matching code make the behaviour of match correct in each case that we consider?

Actually yes. I made the change from `f(x).diff(x) + P_*f(x) - Q_*f(x)**n` to `f(x).diff(x) + P_*f(x) + Q_*f(x)**n` as discussed above and it seems to better match the equations. I suspect this is because all wild symbols are on the LHS in the changed pattern. For any terms on the RHS, it adds a negation to treat them as being on LHS and that creates problems.  After the change, `classify_ode` works great.
```python
In [4]: de = A*(f(x)+f(x).diff(x))-f(x)-f(x).diff(x)-2                                                                            

In [5]: classify_ode(de)                                                                                                          
Out[5]: 
('factorable',
 'separable',
 '1st_exact',
 '1st_linear',
 'Bernoulli',
 'almost_linear',
 '1st_power_series',
 'lie_group',
 'nth_linear_constant_coeff_undetermined_coefficients',
 'nth_linear_constant_coeff_variation_of_parameters',
 'separable_Integral',
 '1st_exact_Integral',
 '1st_linear_Integral',
 'Bernoulli_Integral',
 'almost_linear_Integral',
 'nth_linear_constant_coeff_variation_of_parameters_Integral')

```

**Comment by mostlyaman** [comment]
@oscarbenjamin i have been busy for the last few days and have finally created a PR. A issue for the wrong integral was already created.

**Comment by smichr** [comment]
The following diff is being tested in my 22462 branch. It passes the tests that you already identified.
```diff
diff --git a/sympy/core/operations.py b/sympy/core/operations.py
index cd0e56fa15..02cccf55e1 100644
--- a/sympy/core/operations.py
+++ b/sympy/core/operations.py
@@ -213,6 +213,7 @@ def _matches_commutative(self, expr, repl_dict=None, old=False):
         equivalent.
 
         """
+        from .function import _coeff_isneg
         # make sure expr is Expr if pattern is Expr
         from .expr import Expr
         if isinstance(self, Expr) and not isinstance(expr, Expr):
@@ -254,7 +255,10 @@ def _matches_commutative(self, expr, repl_dict=None, old=False):
                 return None
             newexpr = self._combine_inverse(expr, exact)
             if not old and (expr.is_Add or expr.is_Mul):
-                if newexpr.count_ops() > expr.count_ops():
+                check = newexpr
+                if _coeff_isneg(check):
+                    check = -check
+                if check.count_ops() > expr.count_ops():
                     return None
             newpattern = self._new_rawargs(*wild_part)
             return newpattern.matches(newexpr, repl_dict)
diff --git a/sympy/core/tests/test_match.py b/sympy/core/tests/test_match.py
index 1f30400419..caa25e3c9d 100644
--- a/sympy/core/tests/test_match.py
+++ b/sympy/core/tests/test_match.py
@@ -756,3 +756,11 @@ def test_match_bound():
     assert Sum(x, (x, 1, 2)).match(Sum(y, (y, 1, W))) == {W: 2}
     assert Sum(x, (x, 1, 2)).match(Sum(V, (V, 1, W))) == {W: 2, V:x}
     assert Sum(x, (x, 1, 2)).match(Sum(V, (V, 1, 2))) == {V:x}
+
+
+def test_issue_22462():
+    x, f = symbols('x'), Function('f')
+    n, Q = symbols('n Q', cls=Wild)
+    pattern = -Q*f(x)**n
+    eq = 5*f(x)**2
+    assert pattern.matches(eq) == {n: 2, Q: -5}
diff --git a/sympy/solvers/ode/tests/test_ode.py b/sympy/solvers/ode/tests/test_ode.py
index c53d38d38b..80c7024997 100644
--- a/sympy/solvers/ode/tests/test_ode.py
+++ b/sympy/solvers/ode/tests/test_ode.py
@@ -1059,3 +1059,10 @@ def test_issue_22604():
     assert x1sol == Eq(x1(t), sqrt(3 - sqrt(5))*(sqrt(10) + 5*sqrt(2))*cos(sqrt(2)*t*sqrt(3 - sqrt(5))/2)/20 + \
                        (-5*sqrt(2) + sqrt(10))*sqrt(sqrt(5) + 3)*cos(sqrt(2)*t*sqrt(sqrt(5) + 3)/2)/20)
     assert x2sol == Eq(x2(t), (sqrt(5) + 5)*cos(sqrt(2)*t*sqrt(3 - sqrt(5))/2)/10 + (5 - sqrt(5))*cos(sqrt(2)*t*sqrt(sqrt(5) + 3)/2)/10)
+
+
+def test_issue_22462():
+    for de in [
+            Eq(f(x).diff(x), -20*f(x)**2 - 500*f(x)/7200),
+            Eq(f(x).diff(x), -2*f(x)**2 - 5*f(x)/7)]:
+        assert 'Bernoulli' in classify_ode(de, f(x))
```

**Comment by mostlyaman** [comment]
Yeah, I can understand that could also work. Maybe this is a better alternative solution. 
(I was possibly distracted by the negation operation and trying to fix that instead of considering something like this.)
Should I close the PR if this is the way to go?

**Comment by oscarbenjamin** [comment]
What exactly is the purpose of the `count_ops` check? Can it not just be removed?

I don't see how the output of `count_ops` could possibly be relevant for determining the success of pattern matching.

**Comment by mostlyaman** [comment]
No idea from all the examples I have tested till now.

**Comment by mostlyaman** [comment]
>I don't see how the output of count_ops could possibly be relevant for determining the success of pattern matching.

Here is how I understand this example. I am trying for `-Q*f(x)**2` to match `5*f(x)**2`.
In the _matches_commutative method, 
```python
(Pdb) pprint(self)
    2   
-Q⋅f (x)
(Pdb) pprint(expr)
   2   
5⋅f (x)
(Pdb) print(wild_part, exact_part)
[Q_] [-1, f(x)**2]
```
A new expression is formed using the exact_part list and stored in exact.
```python
(Pdb) pprint(exact)
  2   
-f (x)
```
self._combine_inverse subtracts/divides the expr by the expression stored in `exact` hoping to get cancellations.
```python
(Pdb) pprint(newexpr)   # here newexpr = expr / exact
-5
```
If `exact` is a part of `expr` and is cancelled sucessfully, it should be true that `exact` has less operations than `expr`.
So, I think the comparison between the no of operations between `expr` and `newexpr` is performed as a check to ensure that there has been successful cancellations.
The operation flow is in the following image.
![image](https://user-images.githubusercontent.com/85072876/148366994-9afc54ac-9996-4f26-b322-6ec4bbf33da3.png)

If there had been no cancellations(maybe because the expression formed by exact_part args is not there in expr, not sure about this though), the no of operations in newexpr will be greater than expr because of the `expr/exact` operation and the method returns None.

So, the if condition does have a purpose and probably should be handled better than straight up removing it. (WIP)

**Comment by mostlyaman** [comment]
```diff
diff --git a/sympy/core/operations.py b/sympy/core/operations.py
index cd0e56fa15..50c39f9c28 100644
--- a/sympy/core/operations.py
+++ b/sympy/core/operations.py
@@ -255,7 +255,8 @@ def _matches_commutative(self, expr, repl_dict=None, old=False):
             newexpr = self._combine_inverse(expr, exact)
             if not old and (expr.is_Add or expr.is_Mul):
                 if newexpr.count_ops() > expr.count_ops():
-                    return None
+                    if not newexpr + expr == 0:
+                        return None


             newpattern = self._new_rawargs(*wild_part)
             return newpattern.matches(newexpr, repl_dict)
```
Something like this also works! and it handles the specific case where the negation manipulation happens. Negation causes the newexpr to be negative of expr and we don't want to return None only in that case.

**Comment by mostlyaman** [comment]
@oscarbenjamin @smichr Either the above solution or smichr's solution can be applied. Can you suggest which would be better? IMHO my solution is more specific to the situation which is causing the bug.

**Comment by oscarbenjamin** [comment]
So basically the issue is when `self` is a `Mul` and `exact = -1` then `self._combine_inverse` returns `-self`. If `self` does not have a leading `Rational` coefficient then the `-1` is not absorbed and we go from something like `Mul(a, b)` to `Mul(-1, a, b)` which is considered to have a higher op count although it should probably be considered equivalent in this context.

Thinking this out it seems like `_combine_inverse` should really return something to indicate whether there was successful cancellation rather than the caller trying to infer the cancellation retrospectively. I would have thought that for `Mul` what you want to do here is write the two expressions as dicts of base-exp pairs and compute the cancellation with those dicts e.g.:
```python
In [2]: (x*y**2).as_powers_dict()
Out[2]: defaultdict(int, {x: 1, y: 2})

In [3]: (y**2).as_powers_dict()
Out[3]: defaultdict(int, {y: 2})

In [4]: x.as_powers_dict()
Out[4]: defaultdict(int, {x: 1})
```
We know that `exact` is a subset of the `args` of `self` and we are trying to compute `expr/exact` so that we can match it with `self/exact`. Why do we need to check for cancellation though? We know there is cancellation in `self/exact` which isolates the wild parts of `self`. Why do we care if there is any cancellation in `expr/exact`?

Okay I see it's because we don't want to add new things over to the expression from the pattern. If something in the pattern isn't "in" the expression then we shouldn't introduce it to the expression e.g.:
```python
In [8]: w = Wild('w')

In [9]: (w*x).matches(y)  # don't want to match w = y/x

In [10]: (w*x*z).matches(x**3*z**3) # matches because rhs has x*z
Out[10]: 
⎧    2  2⎫
⎨w: x ⋅z ⎬
⎩        ⎭
```
To me that says that we want `_combine_inverse` to succeed only when the exponents in the powers dict for exact are all `<=` those in `expr` something like this:
```python
def as_coeff_powers_dict(expr):
    coeff, mul = expr.as_coeff_Mul()
    return coeff, mul.as_powers_dict()

def _combine_inverse(lhs, rhs):
    """Computes lhs/rhs but only if rhs is "contained" in lhs"""
    lhs_c, lhs_d = as_coeff_powers_dict(lhs)
    rhs_c, rhs_d = as_coeff_powers_dict(rhs)
    for base, exp in rhs_d.items():
        # Check rhs is "contained" in lhs e.g. x**2 is contained in x**3 but
        # not vice versa
        if (lhs_d[base] < exp) == True:
            return None
        lhs_d[base] -= exp
    args = [lhs_c/rhs_c] + [b**e for b, e in lhs_d.items()]
    return Mul(*args)
```
I guess that special handling for negative/symbolic exponents might also be needed. Perhaps that test should also delegate to `Add._combine_inverse` which could also be implemented using `as_coefficients_dict`.

I have to say I don't really understand the current semantics of `match` which seems to be more implementation defined than based on a consistent principle. For example:
```python
In [23]: (w/x).matches(y*x)

In [24]: (w/x).matches(y*x**2)
Out[24]: 
⎧    3  ⎫
⎨w: x ⋅y⎬
⎩       ⎭
```
The difference between these two is due to the `count_ops` check but I don't really know how to justify that semantically. It seems to me that either both of these should match or neither should. I think that my preference would be for neither.

It would be good to check the behaviour of other pattern matching implementations e.g. `matchpy` to see how sympy's current behaviour compares. The most important thing that is needed is a clear definition somewhere of what things should be expected to match.

**Comment by oscarbenjamin** [comment]
The suggested fixes in the PRs seem to presume that the issue is only about negative numbers but it also applies to Rational and I'm sure many other things e.g.:
```python
In [3]: (2*Q*f(x)**n).matches(5*f(x)**2)
```
There's just no way that `count_ops` should play any role in this:
```python
In [4]: (2*x).count_ops()
Out[4]: 1

In [5]: (2*x/3).count_ops()
Out[5]: 2

In [6]: (-2*x/3).count_ops()
Out[6]: 3
```

**Comment by oscarbenjamin** [comment]
The bug has not been fully fixed e.g. after #22799 we still have
```python
In [6]:     x, f = symbols('x'), Function('f')
   ...:     n, Q = symbols('n Q', cls=Wild)

In [7]: (2*Q*f(x)**n).matches(5*f(x)**2)  # should match

In [8]: (-Q*f(x)**n).matches(5*f(x)**2)
Out[8]: {Q: -5, n: 2}
```

**Comment by oscarbenjamin** [comment]
Also I don't think that #22799 is really the best way to fix this. It would be better at least to separate the number coefficient from the other args when doing `combine_inverse`.

**Comment by mostlyaman** [comment]
> I have to say I don't really understand the current semantics of match which seems to be more implementation defined than based on a consistent principle.

I agree. SymPy's current matching technique is very inconsistent. I tried some of your examples and changed some bits and it didnt take long to find weird behaviour.
```python
>>> Q = Wild('Q')
>>> n = Wild('n')
>>> print((2*Q*f(x)**n).matches(5/2*f(x)**2))
{n_: 2, Q_: 1.25000000000000}
#Here i guess it was able to return because the coeff already is a rational
#thus dividing by the coeff of Wild Part does not really caused a increase in count_ops
#count_ops is not really a good choice for use here
>>> print((2*Q*f(x)**n).matches(5*f(x)**2))
None
```
>The suggested fixes in the PRs seem to presume that the issue is only about negative numbers but it also applies to Rational and I'm sure many other things 


As you can see in the example, i dont think it bugs exactly on -ve numbers or Rationals as it was able to match for one containing Rational. match needs a proper robust solution (atleast something which is consistent) and possibly some dataset of expressions should be added to tests as it is expected to work on a very wide scope of inputs. I will look for better approaches as you suggested.

**Comment by oscarbenjamin** [comment]
Looking at some other pattern matching implementations (matchpy and Mathematica) they are a lot stricter about what they will accept. For example if there is any numeric coefficient in the pattern then that coefficient must appear precisely in target expression so `(2*Q*f(x)**n).matches(5*f(x)**2)` wouldn't match and neither would `(-Q*f(x)**n).matches(f(x)**2)`.

Each pattern symbol needs to match to something that is actually present in the target expression and each `func` has to match. There would be no match for any of these:
```python
In [10]: w = Wild('w')

In [11]: n = Wild('n')

In [12]: (w**n).matches(x)
Out[12]: {n: 1, w: x}

In [13]: (w/x).matches(x)
Out[13]: 
⎧    2⎫
⎨w: x ⎬
⎩     ⎭

In [14]: (w*x).matches(x)
Out[14]: {w: 1}

In [15]: (w+x).matches(x)
Out[15]: {w: 0}

In [17]: (-w*x).matches(x)
Out[17]: {w: -1}
```
I expect this stricter matching is needed to be able to implement efficient algorithms with predictable behaviour.
